bsp.6
.TH bsp 6 "14 February 1999"

.SH NAME
bsp \- node builder for WAD files

.SH SYNOPSIS
.B bsp [options]
.I input.wad
.B [-o
.I output.wad
.B ]

.SH DISCLAIMER
This program is furnished on an AS IS basis for the fun of it.
id Software does not support this program.

.SH DESCRIPTION
.B Bsp
builds the node and blockmap resources for a level map in
a WAD file. Level maps created with level editors and level
compilers won't run without this resources build by a node builder.
The program supports multilevel building, means it automatically
builds the resources for all levels found in the input WAD file.
.LP
If the program is used without the \-o option it will create an output
WAD file with the name tmp.wad.
.LP
.B Bsp
has some special effect features:
.LP
If a linedef has a sector tag >= 900, then it is treated as "precious"
and will not be split unless absolutely necessary. This is good to use
around borders of deep water, invisible stairs, etc.
.LP
Furthermore, just for a grin, if the lindef's tag is 999, then the
sidedef's x-offset set an angle adjustment in degrees to be
applied \-\- you can look straight at a wall, but it might come
right at you on both sides and "stretch".
.LP
.B Bsp
supports HOM-free transparent doors. Simply make the sector
referenced by the doortracks have a sector tag of >= 900. No need
to remember sector numbers and type them in on a command line \-\-
just use any sector tag >= 900 to permanently mark the sector
special.

.SH OPTIONS
.TP
\fB\-factor\fR \fIn\fR
Changes the cost assigned to seg splits. Factor \fIn\fR can be any positive
integer, and the larger the factor, the more costly seg splits are.
A factor of 17 is the default and behaves just like earlier versions
of
.B bsp.
Changing the factor can sometimes prevent visplane overflows in WAD's.
But then again, some WAD's have just too many visplane overflows,
experiment with the \-factor option before giving up.
.TP
.B \-vp
Attempts to prevent visplane overflows without loss of detail. Using the
option causes node line picker to choose node lines in a way that is
empirically known to reduce or eliminate visplane overflows.
.TP
.B \-vpwarn
Warns about potential visplane overflows.
.TP
.B \-vpmark
Marks visplane overflows with player starts. It's a good idea to write
to a separate output file when using this option! One can use an editor
then, play each alternate player start, and see if each one causes
visplane overflows, and then remove it. Usually no-clipping must be used,
at least to step out of walls, since it inserts player starts without
regards to collisions, and it uses walls as a basis for visplane
calculation.
.TP
\fB\-thold\fR \fIn\fR
Threshold for visplane overflow. May be used to report
more potential visplane overflows if the \-vpwarn or \-vpmark options
estimate too few visplanes. Default threshold \fIn\fR is 128.
.TP
.B \-noreject
Does not clobber already existing reject map.
.TP
.B \-noprog
Does not show program progress, for usage from GUI's and such.

.SH SUPPORTED GAMES
Doom, Ultimate Doom, Doom][ and Final Doom.

.SH SEE ALSO
The Unofficial Doom Specs by Matthew S Fell.
.br
.B idbsp (6)
.br
.B warm (6)
.br
.B wreject (6)

.SH BUGS/ERRORS
Will only work on little endian systems.
.br
This program will \fBNOT\fR built a reject map, it will build a zero-ed one,
where every sector is visible from any other sector, or with the appropriate
option it won't touch an existing reject map. So if this node builder is used
the PWAD reject resource should be build with other tools.

.SH AUTHOR
.B Bsp
was written by Colin Reed and Lee Killough (killough@rsn.hp.com).
Andre  Majorel  did the Linux port, Udo Munk integrated the sources
into the WAD tool collection and minimal changes were made to compile
on other UNIX systems besides Linux.


NEWS
-------------------------------------------------------------------------------
Changes in 3.0x:

Lee Killough 9/98:

Fixed many cases of slime lines by going back to using slightly slower BSP21x
splitting algorithm.

Removed approximate visplane overflow detection code, as being too inaccurate
to be useful. If visplane overflows need to be detected, then modifying the
Doom source code to display them exactly in real-time would be a better idea.
The visplane counting code in BSP22x and BSP23x was never 100% accurate, but
was about as close as I could make it based solely on empirical evidence.

NOTE: this change does not remove the -vp option, which attempts to reduce
the chances of visplane overflows by using a special node line heuristic, and
which is the more useful of the visplane-related options. This change only
removes the -vpwarn and -vpmark options, which attempted to detect where
visplane overflows occurred, but which were too crude. Exact detection of
visplane overflows is unnecessary to remove them, because techniques are
available which reduce the probability of visplane overflows without having
to know whether they actually occur.

-------------------------------------------------------------------------------
Changes in 2.3x:

Lee Killough 2/98:

Fixed a minor bug which caused textures to bleed sometimes.
Thanks to Jonathan Campbell (joncamp@tiac.net) for finding it.

Switched from djgpp v1.12 to v2.01.

Linux builds now supported (thanks to Andre Majorel for the port).

Possible future enhancement:

>64K blockmap support (requires changes in Doom source code as well).

-------------------------------------------------------------------------------
Changes in 2.2x:

Lee Killough 10/97:

Added -vp option to reduce the chances of visplane overflows, without loss
of detail. Using -vp causes node line picker to choose node lines in a way
that is empirically known to reduce or eliminate visplane overflows (see
visplane.txt). No guarantees, though!!!

Added -vpwarn option to warn about possible visplane overflows (see
visplane.txt).

-vpmark option also added, which leaves permanent marks in the form of player
1 starts. It's a good idea to write to a separate output file when using this
option!!!  You can then use an editor, play each alternate player start, and
see if each one causes visplane overflows, and then remove it. Usually no-
clipping must be used, at least to step out from the wall, since it inserts
player starts without regard to collisions, and it uses walls as a basis for
visplane calculations.

Added -noreject option, to allow building nodes without messing up an
already-built reject map.

Added -thold option to change the visplane threshold. May be used to report
more potential visplane overflows if the -vpwarn or -vpmark approximations
estimate too few visplanes. Default threshold is 128.

Changed the way level markers in the lump directory are handled, to
accomodate DCK3.61, EdMap, and perhaps other editors. Some editors, such
as DCK3.61, do not always set the length field in a level marker to 0, and
EdMap requires that the offset field in a level marker be equal to the
offset of the first resource.

-------------------------------------------------------------------------------
Changes in 2.1x:

Lee Killough 06/97

Fixed design bug which caused invisible barriers or disappearing Things.
Problem was in how segs incident to a partitioning node line were
partitioned -- they were partitioned according to sidedef, rather than
vertex, orientation w.r.t. the node line. Flipping the affected seg's
linedef usually fixed the problem. Now an incident seg is properly
partitioned either on the left or right side of a node line depending
on whether the two segs are in the same direction or not, not whether
they are both flipped or not.

Added special effects features:

If a linedef has a sector tag >= 900, then it is treated as "precious"
and will not be split unless absolutely necessary. This is good to use
around borders of deep water, invisible stairs, etc.

Furthermore, just for grins, if the linedef's tag is 999, then the sidedefs'
x-offsets set an angle adjustment in degrees to be applied -- you can look
straight at a wall, but it might come right at you on both sides and
"stretch".

-----------------------------------------------------------

New to 2.0x:

Multilevel support!!!  Automatically builds all levels in wad.

Does not strip textures or other lumps from wads anymore!!!

Supports HOM-free transparent doors:

  Simply make the sector referenced by the doortracks have a
  sector tag of >= 900. No need to remember sector numbers
  and type them in on a command line -- just use any sector
  tag >= 900 to permanently mark the sector special.

  See TRANSDOR.WAD for an example of sector tags in the 900's
  being used to create HOM-free transparent door effects.

-----------------------------------------------------------

Changes from 1.4x to 1.5x:

Replaced GO32.EXE with newer version, to support Windows
and DPMI.

-----------------------------------------------------------

Changes from 1.3x to 1.4x:

Lee Killough:

Fixed bug which caused BSP to "do nothing" on systems where
GO32.EXE was not in the user's path.

Added -factor option to allow changing the heuristic used by the
node picker. Factor can be any positive integer, and the larger
the factor, the more costly seg splits are. A factor of 17 is the
default and behaves just like earlier versions of BSP. Changing the
factor can sometimes prevent visplane overflows in wads. But then
again, some wads have just too many visible 2s linedefs for the
node builder to make a difference. If you have visplane overflows,
experiment with the -factor option before giving up.

-----------------------------------------------------------

Changes from 1.2x to 1.3x:

New version due to performance enhancements by Lee Killough
(killough@rsn.hp.com).

Runs at least 3 times as fast as before, due to changes in PickNode.

bsp.c
/*- BSP.C -------------------------------------------------------------------*

 Node builder for DOOM levels (c) 1998 Colin Reed, version 3.0 (dos extended)

 Performance increased 200% over 1.2x

 Many thanks to Mark Harrison for finding a bug in 1.1 which caused some
 texture align problems when a flipped SEG was split.

 Credit to:-

 Raphael Quinet (A very small amount of code has been borrowed from DEU).

 Matt Fell for the doom specs.

 Lee Killough for performance tuning, support for multilevel wads, special
 effects, and wads with lumps besides levels.

 Also, the original idea for some of the techniques where also taken from the
 comment at the bottom of OBJECTS.C in DEU, and the doc by Matt Fell about
 the nodes.

 Use this code for your own editors, but please credit me.

*---------------------------------------------------------------------------*/

#include "structs.h"
#include "bsp.h"

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#include <errno.h>

/*- Global Vars ------------------------------------------------------------*/

static FILE *outfile;
static char *testwad;
char *outwad;

static struct directory *direc = NULL;

struct Seg *(*PickNode)(struct Seg *)=PickNode_traditional;
static int visplane;
static int noreject;

static unsigned char pcnt;

static struct lumplist *lumplist,*current_level;

static struct wad_header wad;

/*- Prototypes -------------------------------------------------------------*/

static FILE *infile;

/* fcopy - function to completely copy one stream to another */

void fcopy(FILE* in, FILE* out)
{
	char buf[1024];
	int rb;
	while ((rb = fread(buf, 1, sizeof buf, in)) > 0) {
		fwrite(buf, 1, rb, out);
	}
}

/*--------------------------------------------------------------------------*/

void progress()
{
	if((verbosity > 1) && !((++pcnt)&31))
		Verbose("%c\b","/-\\|"[((pcnt)>>5)&3]);
}

/*- get the directory from a wad file --------------------------------------*/
/* rewritten by Lee Killough to support multiple levels and extra lumps */

static int OpenWadFile(char *filename)
{
 long i;
 register struct directory *dir;
 struct lumplist *levelp=NULL;
 int levels = 0;

 if (!(infile = fopen(filename,"rb")))
   ProgError("Error: Cannot open WAD file %s", filename);

#if defined(HAVE_TMPFILE)
 if (fseek(infile,0,SEEK_SET) == -1) {
   FILE* intmp;
   if (errno != ESPIPE) {
     perror("fseek"); exit(errno);
   }
   if (!(intmp = tmpfile())) {
     perror("tmpfile"); 
     ProgError("Input was not seekable and failed to create temp file");
   }
   Verbose("Copying piped input to temporary file\n");
   fcopy(infile,intmp);
   fclose(infile);
   rewind(infile = intmp);
 }
#endif

 if (fread(&wad,1,sizeof(wad),infile)!=sizeof(wad) ||
     (wad.type[0]!='I' && wad.type[0]!='P') || wad.type[1]!='W'
     || wad.type[2]!='A' || wad.type[3]!='D')
   ProgError("%s does not appear to be a wad file -- bad magic", filename);

 Verbose("Opened %cWAD file : %s. %lu dir entries at 0x%lX.\n",
	wad.type[0],filename,wad.num_entries,wad.dir_start);

 direc = GetMemory(sizeof(struct directory) * wad.num_entries);

 fseek(infile,wad.dir_start,0);
 fread(direc,sizeof(struct directory),wad.num_entries,infile);

 current_level=NULL;
 dir = direc;
 for (i=wad.num_entries;i--;dir++)
  {
   register struct lumplist *l=GetMemory(sizeof(*l));
   l->dir=dir;
   l->data=NULL;
   l->next=NULL;
   l->islevel=0;

   if ((dir->name[0]=='E' && dir->name[2]=='M' &&
        dir->name[1]>='1' && dir->name[1]<='9' &&
        dir->name[3]>='1' && dir->name[3]<='9' &&
       (!dir->name[4] || (dir->name[4]>='0' &&
                          dir->name[4]<='9' &&
                         !dir->name[5])  )) || (
        dir->name[0]=='M' && dir->name[1]=='A' &&
        dir->name[2]=='P' && !dir->name[5] &&
        dir->name[3]>='0' && dir->name[3]<='9' &&
        dir->name[4]>='0' && dir->name[4]<='9'))
    {
     dir->length=0;
     l->islevel=1;                                 /* Begin new level */
     levels++;
    }
   else
    {
     if (levelp && (!strncmp(dir->name,"SEGS",8) ||
                    !strncmp(dir->name,"SSECTORS",8) ||
                    !strncmp(dir->name,"NODES",8) ||
                    !strncmp(dir->name,"BLOCKMAP",8) ||
                    (!noreject && !strncmp(dir->name,"REJECT",8)) ||
                    FindDir(dir->name)))
       continue;  /* Ignore these since we're rebuilding them anyway */

     if (levelp && FindDir(dir->name))
      {
       Verbose("Warning: Duplicate entry \"%-8.8s\" ignored in %-.8s\n",
         dir->name, current_level->dir->name);
       continue;
      }

     if (levelp && (!strncmp(dir->name,"THINGS",8) ||
         !strncmp(dir->name,"LINEDEFS",8) ||
         !strncmp(dir->name,"SIDEDEFS",8) ||
         !strncmp(dir->name,"VERTEXES",8) ||
         !strncmp(dir->name,"SECTORS",8) ||
         (noreject && !strncmp(dir->name,"REJECT",8))))
      {                       /* Store in current level */
       levelp->next=l;
       levelp=l;
       continue;
      }
    }

    /* Mark end of level and advance one main lump entry */

   if (levelp && !(current_level->data=current_level->next))
     current_level->islevel=0;  /* Ignore oddly formed levels */

   levelp = l->islevel ? l : NULL;

   if (current_level)
     current_level->next=l;
   else
     lumplist=l;
   current_level=l;
  }

 if (levelp)
  {
   if (!(current_level->data=current_level->next))
     current_level->islevel=0;  /* Ignore oddly formed levels */
   current_level->next=NULL;
  }

  return levels;
}

/*- find the pointer to a resource -----------------------*/

struct lumplist *FindDir(const char *name)
{
 struct lumplist *l = current_level;
 while (l && strncmp(l->dir->name,name,8))
   l=l->next;
 return l;
}

/* ReadLump - read a lump into memory */

void* ReadLump(struct lumplist *l)
{
  struct directory *dir = l->dir;
  if (!l->data && dir->length) {
    l->data=GetMemory(dir->length);
    if (fseek(infile,dir->start,0) ||
        fread(l->data,1,dir->length,infile) != dir->length)
         ProgError("Unable to read wad directory entry \"%-8.8s\" in %-.8s\n",
                dir->name, current_level->dir->name);
  }
  return l->data;
}

/* Add a lump to current level
   by Lee Killough */

void add_lump(const char *name, void *data, size_t length)
{
 struct lumplist *l;
 for (l=current_level;l;l=l->next)
   if (!strncmp(name,l->dir->name,8))
     break;
 if (!l)
  {
   l=current_level;
   while (l->next)
     l=l->next;
   l->next = GetMemory(sizeof(*l));
   l = l->next;
   l->next=NULL;
   l->dir = GetMemory(sizeof(struct directory));
   strncpy(l->dir->name,name,8);
  }
 l->dir->length=length;
 l->islevel=0;
 l->data=data;
}

static struct directory write_lump(struct lumplist *lump)
{
 ReadLump(lump); /* cph - fetch into memory if not there already */
 if ((lump->dir->start = ftell(outfile)) == -1 || (lump->dir->length &&
   fwrite(lump->data, 1, lump->dir->length, outfile) != lump->dir->length))
   ProgError("Failure writing %-.8s\n", lump->dir->name);
 if (!lump->islevel) { free(lump->data); lump->data = NULL; }
 return *lump->dir;
}

static void sortlump(struct lumplist **link)
{
 static const char *const lumps[10]={"THINGS", "LINEDEFS", "SIDEDEFS",
  "VERTEXES", "SEGS", "SSECTORS", "NODES", "SECTORS", "REJECT", "BLOCKMAP"};
 int i=sizeof(lumps)/sizeof(*lumps)-1;
 struct lumplist **l;
 do
   for (l=link;*l;l=&(*l)->next)
     if (!strncmp(lumps[i],(*l)->dir->name,8))
      {
       struct lumplist *t=(*l)->next;
       (*l)->next=*link;
       *link=*l;
       *l=t;
       break;
      }
 while (--i>=0);
}


void usage(void)
{
 printf("\nThis Node builder was created from the basic theory stated in DEU5 (OBJECTS.C)\n"
        "\nCredits should go to :-\n"
        "Matt Fell      (msfell@aol.com) for the Doom Specs.\n"
        "Raphael Quinet (Raphael.Quinet@eed.ericsson.se) for DEU and the original idea.\n"
        "Mark Harrison  (harrison@lclark.edu) for finding a bug in 1.1x\n"
        "Jim Flynn      (jflynn@pacbell.net) for many good ideas and encouragement.\n"
        "Jan Van der Veken for finding invisible barrier bug.\n"
#ifdef MSDOS
        "\nUsage: BSP [options] input.wad [[-o] <output.wad>]\n"
#else
        "\nUsage: BSP [options] input.wad [-o <output.wad>]\n"
#endif
        "       (If no output.wad is specified, TMP.WAD is written)\n\n"
        "Options:\n\n"
        "  -factor <nnn>  Changes the cost assigned to SEG splits\n"
        "  -vp            Attempts to prevent visplane overflows\n"
        "  -noreject      Does not clobber reject map\n"
	"  -q             Quiet mode (only errors are printed\n"
       );
 exit(1);
}

static int quiet;

static void parse_options(int argc, char *argv[])
{
 static char *fnames[2];
 static const struct {
   const char *option;
   void *var;
   enum {NONE, STRING, INT} arg;
 } tab[]= { {"-vp", &visplane, NONE},
            {"-noreject", &noreject, NONE},
            {"-q", &quiet, NONE},
            {"-factor", &factor, INT},
            {"-o", fnames+1, STRING},
          };
 int nf=0;

 while (--argc)
   if (**++argv=='-')
    {
     int i=sizeof(tab)/sizeof(*tab);
     for (;;)
       if (!strcmp(*argv,tab[--i].option))
        {
         if (tab[i].arg==INT)
           if (--argc)
            {
             char *end;
             *(int *) tab[i].var=strtol(*++argv,&end,0);
             if (*end || factor<0)
               usage();
            }
           else
             usage();
         else
	   if (tab[i].arg==STRING)
	     if (--argc)
	       *(char **) tab[i].var = *++argv;
	     else
	       usage();
	   else
	     ++*(int *) tab[i].var;
         break;
        }
       else
         if (!i)
          {
           usage();
           break;
          }
    }
   else
#ifdef MSDOS
     if (nf<2)
#else
     if (nf<1)
#endif
       fnames[nf++]=*argv;
     else
       usage();

 testwad = fnames[0];                          /* Get input name*/

 if (!testwad || factor<0)
   usage();

 outwad = fnames[1] ? fnames[1] : "tmp.wad";   /* Get output name*/
}

/*- Main Program -----------------------------------------------------------*/

int main(int argc,char *argv[])
{
 struct lumplist *lump,*l;
 struct directory *newdirec;
 int levels;

 parse_options(argc,argv);

 verbosity = quiet ? 0 : 
#ifdef HAVE_ISATTY
	isatty(STDOUT_FILENO) ? 2 : 1
#else
	2
#endif
	;

 /* Don't buffer output to stdout, people want to see the progress 
  * as it happens */
#ifdef HAVE_ISATTY
 if (isatty(STDOUT_FILENO))
#endif
   setbuf(stdout,NULL);

 if (verbosity)
  Verbose("* Doom BSP node builder ver " VERSION "\n"
	"Copyright (c)	1998 Colin Reed, Lee Killough\n"
	"		2000 Colin Phipps <cph@lxdoom.linuxgames.com>\n\n");

 levels = OpenWadFile(testwad);		/* Opens and reads directory*/

 Verbose("Creating nodes using tunable factor of %d\n",factor);

 if (visplane)
  {
   Verbose("\nTaking special measures to reduce the chances of visplane overflow");
   PickNode=PickNode_visplane;
  }

  if (!(outfile=fopen(outwad,"wb")))
   {
    fputs("Error: Could not open output PWAD file ",stderr);
    perror(outwad);
    exit(1);
   }

 /* Allocate space for existing lumps plus some extra for each level */
 newdirec = GetMemory((wad.num_entries + 10*levels)*sizeof(struct directory));

 wad.num_entries = 0;

 fwrite("xxxxxxxxxxxx",sizeof wad,1,outfile);

 for (lump=lumplist; lump; lump=lump->next) {
  newdirec[wad.num_entries++]=write_lump(lump);
  if (lump->islevel)
    {
     char current_level_name[9];
     strncpy(current_level_name,lump->dir->name,8);
     current_level_name[8] = 0;
     DoLevel(current_level_name, current_level = lump->data);
     sortlump((struct lumplist **) &lump->data);
     for (l=lump->data; l; l=l->next)
       newdirec[wad.num_entries++]=write_lump(l);
   }  /* if (lump->islevel) */
 }

 if ((wad.dir_start = ftell(outfile)) == -1 ||
    fwrite(newdirec,sizeof(struct directory),wad.num_entries,outfile)!=wad.num_entries)
    ProgError("Failure writing lump directory");

 if (fseek(outfile, 0, SEEK_SET) || fwrite(&wad,1,12,outfile)!=12)
    ProgError("Failure writing wad header");

  fclose(outfile);

  Verbose("\nSaved WAD as %s\n",outwad);

 return 0;
}

/*- end of file ------------------------------------------------------------*/
